src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,217 @@
-          log "Compile time errors detected. Shutting down..."
-          exit 1
+    private def run_watcher(watch_object)
+      watch_object.each do |key, value|
+        next if key == "client"
+        files = value["files"]
+        commands = value["commands"]
+        if key != "server"
+          @notify_counter += 1
         end
+        spawn watcher(key, files, commands)
       end
+      @notify_counter_channel.send @notify_counter
+      @notify_counter = @notify_counter_channel.receive
+      sleep
     end
 
-    private def scan_files
+    private def watcher(key, files, commands)
+      if key != "server"
+        @notify_channel.receive
+      end
+      if files.empty?
+        commands.each do |command|
+          run_command(command, key)
+        end
+      else
+        loop do
+          scan_files(key, files, commands)
+          @watch_running = true if key == "server"
+          sleep 1
+        end
+      end
+    end
+
+    private def scan_files(key, files, commands)
       file_counter = 0
-      Dir.glob(files) do |file|
-        timestamp = get_timestamp(file)
-        if FILE_TIMESTAMPS[file]? != timestamp
-          if @app_running
-            log "File changed: #{file.colorize(:light_gray)}"
-          end
-          FILE_TIMESTAMPS[file] = timestamp
-          file_counter += 1
+      @file_watcher.scan_files(files) do |file|
+        if @watch_running
+          debug "File changed: #{file}"
         end
+        file_counter += 1
       end
       if file_counter > 0
-        log "Watching #{file_counter} files (server reload)..."
-        start_app
+        debug "Watching #{file_counter} #{key} files"
+        kill_processes(key)
+        commands.each do |command|
+          if key == "server" && command == commands.first?
+            run_build_command(command, commands)
+          else
+            run_command(command, key)
+          end
+        end
       end
     end
 
-    private def stop_all_processes
-      log "Terminating app #{project_name}..."
-      @processes.each do |process|
-        process.kill unless process.terminated?
+    private def check_directories
+      Dir.mkdir_p("bin")
+      if !Dir.exists?("lib")
+        error "You need to install dependencies first, execute `shards install`"
+        exit 1
       end
-      processes.clear
     end
 
-    private def create_all_processes
-      process = create_watch_process
-      @processes << process if process.is_a? Process
-      unless @npm_process
-        create_npm_process
-        @npm_process = true
+    private def run_build_command(command, commands)
+      check_directories
+      next_server_commands_range = (1...commands.size)
+      info "Building project #{Helpers.settings.name.colorize(:light_cyan)}"
+      spawn do
+        error_io = IO::Memory.new
+        process = Helpers.run(command, error: error_io)
+        PROCESSES << {process, "server"}
+        loop do
+          if process.terminated?
```

Comment:
I think this loop contains far too much logic all in one place. It may help to refactor to use early-exits instead, but I don't know. It's hard to even see what's going on here, but something like this might be possible:

```crystal

loop do
   sleep 1
   next if process.terminated?
   next if error_io.empty?
   # etc
end
```

Reply:
I added a new method `handle_terminaded_process`:

Now the `loop` looks like this:

```cr
        loop do
          if process.terminated?
            handle_terminaded_process(process, error_io)
            break
          end
          sleep 1
        end
```